src/qs8-igemm/c8-wasmdot.c.in

```diff
@@ -283,20 +283,14 @@ void xnn_${DATATYPE_SPEC}_igemm_minmax${REQUANTIZATION_SPEC}_ukernel_${MR}x${NR}
         $else:
           v128_t vout = wasm_i8x16_narrow_i16x8(vacc0${min(1, MR-1)}x0123, vacc0${min(1, MR-1)}x0123);
       $else:
-        $if MR > 2:
-          v128_t vout0x01234567_1x01234567 = wasm_i8x16_narrow_i16x8(vacc0x01234567, vacc1x01234567);
-          v128_t vout2x01234567_${min(3, MR-1)}x01234567 = wasm_i8x16_narrow_i16x8(vacc2x01234567, vacc${min(3, MR-1)}x01234567);
-        $else:
-          v128_t vout0x01234567_${min(1, MR-1)}x01234567 = wasm_i8x16_narrow_i16x8(vacc0x01234567, vacc${min(1, MR-1)}x01234567);
+        $for M in range(0, MR, 2):
+          v128_t vout${M}x01234567_${min(M+1, MR-1)}x01234567 = wasm_i8x16_narrow_i16x8(vacc${M}x01234567, vacc${min(M+1, MR-1)}x01234567);
```

Comment:
nit - I'm not a fan of combining rows during quantization.
1. its complicated
2. its often slower
Its a lot simplier to keep 1 vector per row, and if available, using a convert directly from i32 to i8
When combined, the pack functions in avx/avx512 require vpermq or shufb.  For SSE its okay but the benefit of 2 (NR=8) values in a register is small and it impeds your ability to try NR=16, revec or large MR values.
With evex encoding, you'd likely get better performance with MR=8

Reply:
Thanks for the comments!
It's true that this additional combined vectors misleading the revector to pack vacc0x01234567 with vacc0x89ABCDEF which in turn caused inserts from incontinuous loads by b4567 and b89AB in the main loop.
Unfortunately WASM SIMD does not define single i32 to i8 convert op. Instead keep 1 vector per row does help, that's why we see obvious speedup with 5x8c8 as the last row is actually converting itself without combining and avoid the misleading revectorization eventually. I will update this patch and see if it helps to further increase MR.

With c8, there's still one additional hadd which is not easy to handle during revectorization and we will generate inserts in the main loop. So I did experiment c4 which does not require hadd and seems work better w/ or w/o revec. I just submitted a new PR https://github.com/google/XNNPACK/pull/7290 for this.

Reply:
Hi Frank, I have updated this patch and fixed the nit. Please take a look again. @fbarchard 
After this change, I did see a slight improvement with higher MR. But due to current revectorize cannot handle the additional horizontal add for c8, the boost is still lower than c4.